File: fizzbuzz/ruby/FizzBuzz.rb
#!/usr/bin/env ruby

require 'minitest/autorun'

describe 'FizzBuzzCalc' do

  it 'requires a single argument' do
    method(:fizz_buzz_calc).arity.must_equal 1
  end

  it 'returns 1 if passed a 1' do
    fizz_buzz_calc(1).must_equal 1
  end

  it 'returns 2 if passed a 2' do
    fizz_buzz_calc(2).must_equal 2
  end

  it 'returns "Fizz" if passed 3' do
    fizz_buzz_calc(3).must_equal 'Fizz'
  end

  it 'return 4 if passed 4' do
    fizz_buzz_calc(4).must_equal 4
  end

  it 'returns "Buzz" if passed 5' do
    fizz_buzz_calc(5).must_equal 'Buzz'
  end

  it 'return "Fizz" if passed 6' do
    fizz_buzz_calc(6).must_equal 'Fizz'
  end

  it 'returns "Buzz" if passed 10' do
    fizz_buzz_calc(10).must_equal 'Buzz'
  end

  it 'returns 2 if passed 2 (from cache)' do
    fizz_buzz_calc(2).must_equal 2
  end

  it 'returns "FizzBuzz" if passed 15' do
    fizz_buzz_calc(15).must_equal 'FizzBuzz'
  end

  it 'returns "FizzBuzz" if passed 30' do
    fizz_buzz_calc(30).must_equal 'FizzBuzz'
  end

end

def fizz_buzz_calc(number)
  @cache ||= {}
  output = number

  if @cache.has_key? number
    output = @cache.fetch(number)
  else
    is_divisible_by_three = (number % 3 == 0)
    is_divisible_by_five  = (number % 5 == 0)

    if is_divisible_by_three && is_divisible_by_five
      output = 'FizzBuzz'
    elsif is_divisible_by_five
      output = 'Buzz'
    elsif is_divisible_by_three
      output = 'Fizz'
    end
    @cache.store(number, output)
  end
  output
end

File: fizzbuzz/javascript/fizzbuzz.js
(function(){

  var assert = require('assert');

  describe('FizzBuzz', function(){
    it('should require a single argument', function(){
      assert.equal(1, fizzBuzzCalc.length);
    });

    it('should return a 1 when passed a 1', function(){
      assert.equal(1, fizzBuzzCalc(1));
    });

    it('should return a 2 when passed a 2', function(){
      assert.equal(2, fizzBuzzCalc(2));
    });

    it('should return a "Fizz" when passed a 3', function(){
      assert.equal("Fizz", fizzBuzzCalc(3));
    });

    it('should return a 4 when passed a 4', function(){
      assert.equal(4, fizzBuzzCalc(4));
    });

    it('should return "Buzz" when passed a 5', function(){
      assert.equal("Buzz", fizzBuzzCalc(5));
    });

    it('should return "Fizz" when passed a 6', function(){
      assert.equal("Fizz", fizzBuzzCalc(6));
    });

    it('should return 7 when passed a 7', function(){
      assert.equal(7, fizzBuzzCalc(7));
    });

    it('should return "Buzz" when passed a 10', function(){
      assert.equal("Buzz", fizzBuzzCalc(10));
    });

    it('should return "FizzBuzz" when passed a 15', function(){
      assert.equal("FizzBuzz", fizzBuzzCalc(15));
    });

    it('should return "FizzBuzz" when passed a 30', function(){
      assert.equal("FizzBuzz", fizzBuzzCalc(30));
    });

  });

  function fizzBuzzCalc(number){
    var output = number,
        isDivisibleByThree = (number % 3 == 0),
        isDivisibleByFive = (number % 5 == 0);

    if(isDivisibleByThree && isDivisibleByFive){
      output = 'FizzBuzz';
    } else if (isDivisibleByFive){
      output = 'Buzz';
    } else if (isDivisibleByThree){
      output = 'Fizz';
    }

    return output;
  }

})();

